File: bindings/cpp/test/test_feed.cpp
// Change feed FEED.* (contract §6 "FEED replay").
#include "backends.hpp"
#include "harness.hpp"
#include "kevy/client.hpp"

using namespace kevy;
using namespace kevy::test;

KEVY_TEST(feed_replay_remote) {
  if (!server_available()) return;
  auto srv = spawn_server_config("[feed]\nenabled = true\n", {"--threads", "1"});
  Client c = Client::connect(srv->url());

  CHECK(c.feed_shards() >= 1);
  auto [gen, off] = c.feed_tail(0);

  c.set("fk1", "v1");
  c.set("fk2", "v2");

  FeedBatch batch = c.feed_read(0, gen, off, std::nullopt, keys({}));
  CHECK(batch.frames.size() >= 2);
  bool saw_set = false;
  for (const auto& f : batch.frames)
    if (!f.argv.empty() && f.argv[0] == "SET") saw_set = true;
  CHECK(saw_set);

  // Resume from the returned cursor: caught up → empty batch.
  FeedBatch next = c.feed_read(0, batch.generation, batch.next_offset, std::nullopt, keys({}));
  CHECK_EQ(next.frames.size(), size_t(0));
}

KEVY_TEST(feed_stale_cursor_resync) {
  if (!server_available()) return;
  auto srv = spawn_server_config("[feed]\nenabled = true\n", {"--threads", "1"});
  Client c = Client::connect(srv->url());
  // A wildly-stale generation surfaces a FEEDRESYNC protocol error.
  try {
    c.feed_read(0, 999999, 0, std::nullopt, keys({}));
    // server may not treat it as unservable — acceptable, no assertion
  } catch (const KevyError& e) {
    CHECK(e.kind() == ErrorKind::Protocol);
  }
}

KEVY_TEST(feed_embedded_unsupported) {
  Client c = Client::connect(unique_mem_url());
  CHECK_EQ(c.feed_shards(), int64_t(1));  // embedded: always 1
  CHECK_THROWS_KIND(c.feed_tail(1), ErrorKind::InvalidInput);   // non-zero shard
  CHECK_THROWS_KIND(c.feed_tail(0), ErrorKind::Unsupported);    // feed disabled
}

File: bindings/python/tests/test_feed.py
"""Change feed FEED.* (contract §3.10, §6)."""

import kevy
import pytest
from conftest import spawn_server


@pytest.fixture(scope="module")
def feed_server():
    srv = spawn_server("--threads", "1", config="[feed]\nenabled = true\n")
    yield srv
    srv.stop()


def test_feed_replay_and_resume(feed_server):
    c = kevy.connect(feed_server.url)
    assert c.feed_shards() >= 1
    gen, off = c.feed_tail(0)

    c.set("fk1", "v1")
    c.set("fk2", "v2")

    batch = c.feed_read(0, gen, off)
    assert len(batch.frames) >= 2
    assert any(f.argv and f.argv[0] == b"SET" for f in batch.frames)

    # Resume from the returned cursor: caught up → empty batch.
    nxt = c.feed_read(0, batch.generation, batch.next_offset)
    assert len(nxt.frames) == 0
    c.close()


def test_stale_cursor_resync(feed_server):
    c = kevy.connect(feed_server.url)
    try:
        c.feed_read(0, 999999, 0)  # wildly-stale generation
    except kevy.ProtocolError:
        pass  # FEEDRESYNC / unservable cursor → Protocol error
    except kevy.KevyError:
        pass
    c.close()


def test_embedded_feed_unsupported():
    c = kevy.connect("mem://feed-bus")
    assert c.feed_shards() == 1
    with pytest.raises(kevy.InvalidInputError):
        c.feed_tail(1)  # non-zero shard
    with pytest.raises(kevy.UnsupportedError):
        c.feed_tail(0)  # feed not enabled on this port's embedded connect
    c.close()
